src/valid/handles.rs

```diff
@@ -0,0 +1,337 @@
+                    | &crate::Expression::AccessIndex { base, .. } => {
+                        this_handle.check_dep(base)?;
+                    }
+                    &crate::Expression::Constant(constant) => {
+                        validate_constant(constant)?;
+                    }
+                    &crate::Expression::Splat { value, .. } => {
+                        this_handle.check_dep(value)?;
+                    }
+                    &crate::Expression::Swizzle { vector, .. } => {
+                        this_handle.check_dep(vector)?;
+                    }
+                    &crate::Expression::Compose { ty, ref components } => {
+                        validate_type(ty)?;
+                        this_handle.check_dep_iter(components.iter().copied())?;
+                    }
+                    &crate::Expression::FunctionArgument(_arg_idx) => (),
+                    &crate::Expression::GlobalVariable(global_variable) => {
+                        global_variable.check_valid_for(global_variables)?;
+                    }
+                    &crate::Expression::LocalVariable(local_variable) => {
+                        local_variable.check_valid_for(local_variables)?;
+                    }
+                    &crate::Expression::Load { pointer } => {
+                        this_handle.check_dep(pointer)?;
+                    }
+                    &crate::Expression::ImageSample {
+                        image,
+                        sampler,
+                        gather: _,
+                        coordinate,
+                        array_index,
+                        offset,
+                        level: _,
+                        depth_ref,
+                    } => {
+                        if let Some(offset) = offset {
+                            validate_constant(offset)?;
+                        }
+
+                        this_handle
+                            .check_dep(image)?
+                            .check_dep(sampler)?
+                            .check_dep(coordinate)?
+                            .check_dep_opt(array_index)?
+                            .check_dep_opt(depth_ref)?;
+                    }
+                    &crate::Expression::ImageLoad {
+                        image,
+                        coordinate,
+                        array_index,
+                        sample,
+                        level,
+                    } => {
+                        this_handle
+                            .check_dep(image)?
+                            .check_dep(coordinate)?
+                            .check_dep_opt(array_index)?
+                            .check_dep_opt(sample)?
+                            .check_dep_opt(level)?;
+                    }
+                    &crate::Expression::ImageQuery { image, query } => {
+                        this_handle.check_dep(image)?.check_dep_opt(match query {
```

Comment:
The way this turned out, I think it'd just be clearer to have the `match` on the outside, and call `check_dep_opt` from the `ImageQuery::Size` arm. Listing out all the variant is good, though.

Reply:
Updated as suggested with 555f4207.